pkg/cloud/externalconn/record.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package externalconn
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/cloud/externalconn/connectionpb"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/security/username"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+)
+
+// externalConnectionRecord is a reflective representation of a row in a
+// system.external_connections table.
+//
+// Each field in this struct has a tag specifying the column name in the
+// system.external_connections table containing the data for the field. Do not
+// manipulate these fields directly, use methods in the ExternalConnection.
+type externalConnectionRecord struct {
+	ConnectionName    string                         `col:"connection_name"`
+	ConnectionType    string                         `col:"connection_type"`
+	ConnectionDetails connectionpb.ConnectionDetails `col:"connection_details"`
+}
+
+// ExternalConnection is a representation of an External Connection.
+//
+// This struct can marshal/unmarshal changes made to the underlying
+// system.external_connections table.
+type ExternalConnection struct {
+	// The "record" for this external connection.  Do not access this field
+	// directly (except in tests); Use Get/Set methods on ExternalConnection
+	// instead.
+	rec externalConnectionRecord
+
+	// Set of changes to this external connection that need to be persisted.
+	dirty map[string]struct{}
+}
+
+// NewExternalConnection creates and initializes an ExternalConnection.
+func NewExternalConnection() *ExternalConnection {
+	return &ExternalConnection{
+		dirty: make(map[string]struct{}),
+	}
+}
+
+// SetConnectionName updates the connection name.
+func (e *ExternalConnection) SetConnectionName(name string) {
+	e.rec.ConnectionName = name
+	e.markDirty("connection_name")
+}
+
+// SetConnectionType updates the connection_type.
+func (e *ExternalConnection) SetConnectionType(connectionType string) {
```

Comment:
Hmm so in my head connection type and connection details were decoupled concepts. In our implementation each connection detail decides what type it belongs to. Maybe we can have `SetConnectionType` take in the connection details instead so that it always uses the type field on the details?

Reply:
I think in my head, I wouldn't _have_ a connection type at all on the proto. I would keep an enum in a related go file, and write a custom accessor on type for users. Connection type is purely a function of the Connection details, so it seems like the wrong model to maintain an entire field for it.

Reply:
Discussed offline, `ConnectionProvider` is unused today and we might not even need it when we teach BACKUP, IMPORT etc to Dial External Connections. So I'm deleting the field and the related enum.